gordon_janitor_gcp/gce_authority.py

```diff
@@ -0,0 +1,181 @@
+        self.rrset_channel = rrset_channel
+        self.kwargs = kwargs
+        self._loop = asyncio.get_event_loop()
+        self.session = aiohttp.ClientSession(loop=self._loop)
+
+    def _get_or_raise(self, key):
+        item = self.config.get(key)
+        if not item:
+            msg = 'gcp.gceauthority configuration missing "{key}" setting'
+            logging.error(msg)
+            raise exceptions.GCPConfigError(msg)
+        return item
+
+    def _get_msg_generator(self):
+        zone = self._get_or_raise('zone')
+        return msg_generators.ARecordMsgGenerator(zone)
+
+    def _get_crm_client(self):
+        keyfile = self._get_or_raise('crm_keyfile')
+        scopes = self.config.get('crm_scopes')
+
+        crm_auth = auth.GoogleAuthClient(
+            keyfile, scopes=scopes, session=self.session)
+        return gcp_clients.GCRMClient(crm_auth, self.session)
+
+    def _get_gce_client(self):
+        tag_blacklist = self.config.get('tag_blacklist', [])
+        _metadata_blacklist = self.config.get('metadata_blacklist', [])
+        metadata_blacklist = [dict([pair]) for pair in _metadata_blacklist]
+        keyfile = self._get_or_raise('gce_keyfile')
+        scopes = self.config.get('gce_scopes')
+
+        gce_auth = auth.GoogleAuthClient(
+            keyfile, scopes=scopes, session=self.session)
+        return gcp_clients.GCEClient(
+            gce_auth,
+            self.session,
+            tag_blacklist=tag_blacklist,
+            metadata_blacklist=metadata_blacklist)
+
+    def build_authority(self):
+        """Coordinate dependencies and create a GCEAuthority object.
+
+        Returns:
+            obj: A configured GCEAuthority object.
+        """
+        crm_client = self._get_crm_client()
+        gce_client = self._get_gce_client()
+        msg_gen = self._get_msg_generator()
+
+        return GCEAuthority(self.config, crm_client, gce_client, msg_gen,
+                            self.rrset_channel, **self.kwargs)
+
+
+def get_authority(config, rrset_channel, **kwargs):
+    builder = GCEAuthorityBuilder(config, rrset_channel, **kwargs)
+    return builder.build_authority()
+
+
+@zope.interface.implementer(interfaces.IAuthority)
+class GCEAuthority:
+    """Gordon-Janitor plugin that gathers "state of truth" GCE instance data.
+
+    Args:
+        config (dict): plugin-specific configuration.
+        crm_client (obj): client used to fetch GCE projects.
+        gce_client (obj): client used to fetch instances for each project.
+        msg_generator (obj): obj used to transform instance data into resource
+            record set messages for the reconciler.
+        rrset_channel (asyncio.Queue): channel to send resource record messages
+            to.
+    """
+    def __init__(self,
+                 config,
+                 crm_client,
+                 gce_client,
+                 msg_generator,
+                 rrset_channel=None,
+                 **kwargs):
```

Comment:
I really _really_ dislike this style :-/ It's not stated in the google python style guide, and a little elusive in pep8, but we should be consistent. It's also weird to read it this way when there is one line per variable definition right after. Maybe @descent3me can weigh in.